Update developer documentation to describe meson builds

betatim opened the pull request:
This updates the developer docs to explain how to build from source with
the new meson build backend.

closes #28748

betatim committed:
Update developer documentation to describe meson builds

This updates the developer docs to explain how to build from source with
the new meson build backend.

betatim committed:
Merge branch 'main' into meson-build-docs

betatim committed:
Remove dedicated editable install section

The default recommendation is to use an editable install so there is no
more need for a dedicated section.

ogrisel commented on doc/developers/advanced_installation.rst:
```diff
@@ -87,7 +87,7 @@ feature, code or documentation improvement).
      python3 -m venv sklearn-env
      source sklearn-env/bin/activate
-     pip install wheel numpy scipy cython
+     pip install wheel numpy scipy cython meson-python
```
Is ninja also installed automatically with pip? It wasn't the case in the past but I haven't tried again recently.

ogrisel commented on doc/developers/advanced_installation.rst:
```diff
@@ -87,7 +87,7 @@ feature, code or documentation improvement).
      python3 -m venv sklearn-env
      source sklearn-env/bin/activate
-     pip install wheel numpy scipy cython
+     pip install wheel numpy scipy cython meson-python
```
Is ninja also installed automatically with pip? It wasn't the case in the past but I haven't tried again recently.

lesteve replied:
Indeed you need `pip install meson-python ninja`. My understanding is that the system `ninja` can be found if it exists, which may be preferred by Meson people. We prefer that everything is contained in the Python environment for convenience so +1 for `pip install ninja meson-python`.

Full disclosure with conda, `conda install meson-python` will install `ninja`, but I I am fine doing `conda install meson-python ninja` anyway for consistence in the instructions between `conda` and `pip`

ogrisel commented on doc/developers/advanced_installation.rst:
```diff
@@ -64,11 +64,11 @@ feature, code or documentation improvement).
    If you installed Python with conda, we recommend to create a dedicated
    `conda environment`_ with all the build dependencies of scikit-learn
-   (namely NumPy_, SciPy_, and Cython_):
+   (namely NumPy_, SciPy_, Cython_ and Meson_):
```
Shouldn't we rather reference and link to meson-python instead? It's easy to find the meson doc from the meson-python page but not necessarily the opposite.

lesteve replied:
On top of this, I think the two references that were removed are more likely to be useful than Meson or meson-python docs:

- `pandas setup doc
  <https://pandas.pydata.org/docs/development/contributing_environment.html#step-3-build-and-install-pandas>`_:
  pandas has a similar setup as ours (no spin or dev.py)
- `scipy Meson doc
  <https://scipy.github.io/devdocs/building/understanding_meson.html>`_ gives
  more background about how Meson works behind the scenes

betatim replied:
I switched to linking to meson-python.

Where would be a good place to add the links Loic? In this bit of the docs we link to the various dependencies and their homepages, so it feels weird to add the two links which are more about background reading.

lesteve replied:
Good question, maybe a note at the end of the "Building from source" section, maybe this should more go in `developers/tips.rst` since this is more suited for scikit-learn maintainers?

betatim replied:
I added a section to the tips page.

ogrisel commented on doc/developers/advanced_installation.rst:
```diff
@@ -552,6 +460,8 @@ the base system and these steps will not be necessary.
 .. _OpenMP: https://en.wikipedia.org/wiki/OpenMP
 .. _Cython: https://cython.org
+.. _Meson: https://mesonbuild.com/
+.. _Ninja: https://ninja-build.org/
```
This does not seem to be used anywhere but I think we should find a way to explain somewhere that meson relies on ninja under the hood to build native extensions.

lesteve replied:
I think the two references that were removed are more likely to be useful than Meson or Ninja docs:

- `pandas setup doc
  <https://pandas.pydata.org/docs/development/contributing_environment.html#step-3-build-and-install-pandas>`_:
  pandas has a similar setup as ours (no spin or dev.py)
- `scipy Meson doc
  <https://scipy.github.io/devdocs/building/understanding_meson.html>`_ gives
  more background about how Meson works behind the scenes

betatim replied:
Ninja was indeed left over from when I thought you didn't have to install it explicitly. Now it is useful again because you have to install it explicitly and I list it as a build time dependency.

ogrisel reviewed (approved):
LGTM, it's already a net improvement.

ogrisel commented:
/cc @lesteve who's expert eyes might spot details.

ogrisel commented:
I gave my +1 without looking at the CI...

lesteve commented on doc/developers/advanced_installation.rst:
```diff
@@ -111,14 +113,14 @@ feature, code or documentation improvement).
-    future).
+    Note `--config-settings editable-verbose=true` is recommended to avoid surprises
+    when you `import sklearn`. meson-python implements editable installs by recompiling
+    when executing `import sklearn`. Even changing Python files involves copying files
```
I am not so sure anymore about "Even changing Python files involves copying files to the Meson build directory", I think it should be removed

betatim replied:
Removed it.

betatim committed:
Explicitly install ninja

betatim committed:
Switch to using the meson `pip install` command

The command to install with meson is slightly different, this commit
replaces all occurences of the old version.

betatim committed:
Remove parallel build section

Meson has a different/default way of building extensions in parallel, so
we do not need this section anymore.

betatim committed:
Link to meson-python instead of generic meson

betatim committed:
Fix cross-references within the docs

betatim commented:
> we need to re-add the `editable_mode` label to the section to avoid breaking URLs

I think the only place linking to `editable_mode` was a line in the "install from source" section. So I removed that because it now was a circular reference.

betatim committed:
Remove unused reference, make meson note more precise

betatim committed:
Remove cython recompile note

Not needed anymore with meson

betatim commented:
> Also just curious, this PR is still in draft, is this because you have other things you still want to do?

Not really, mostly just hadn't gotten around to move it to "not a draft".

Fixed the `pip install` you pointed to.

betatim committed:
Add a section on meson background information

lesteve committed:
Tweak note

lesteve reviewed (approved):
I pushed a tweak about the meson-python rebuilding on import. LGTM, thanks a lot for tackling this!

lesteve commented:
OK let's merge this, I think this can be improved in a further PR, if we ever feel the need. Thanks a lot!
